src/Headroom.js

```diff
@@ -95,378 +14,206 @@ function normalizeTolerance(t) {
  * @param {Object} options options for the widget
  */
 function Headroom(elem, options) {
-  options = extend(options, Headroom.options);
+  options = options || {};
+  Object.assign(this, Headroom.options, options);
+  Object.assign(this.classes, Headroom.options.classes, options.classes);
 
-  this.lastKnownScrollY = 0;
   this.elem = elem;
-  this.tolerance = normalizeTolerance(options.tolerance);
-  this.classes = options.classes;
-  this.offset = options.offset;
-  this.scroller = options.scroller;
+  this.tolerance = normalizeTolerance(this.tolerance);
   this.initialised = false;
-  this.onPin = options.onPin;
-  this.onUnpin = options.onUnpin;
-  this.onTop = options.onTop;
-  this.onNotTop = options.onNotTop;
-  this.onBottom = options.onBottom;
-  this.onNotBottom = options.onNotBottom;
   this.frozen = false;
 }
 Headroom.prototype = {
   constructor: Headroom,
 
   /**
-   * Initialises the widget
+   * Start listening to scrolling
+   * @public
    */
   init: function() {
-    if (!Headroom.cutsTheMustard) {
-      return;
-    }
-
-    this.debouncer = new Debouncer(this.update.bind(this));
-    this.elem.classList.add(this.classes.initial);
+    if (Headroom.cutsTheMustard && !this.initialised) {
+      this.addClass("initial");
+      this.initialised = true;
 
-    // defer event registration to handle browser
-    // potentially restoring previous scroll position
-    setTimeout(this.attachEvent.bind(this), 100);
+      // defer event registration to handle browser
+      // potentially restoring previous scroll position
+      setTimeout(
+        function(self) {
+          self.scrollTracker = trackScroll(
+            self.scroller,
+            self.update.bind(self)
+          );
+        },
```

Comment:
I think I'm going to drop this asynchronous behaviour in a future version. I would have done it now, but I wanted to avoid breaking changes. Developers can always defer initialisation themselves if it matters to them.

Reply:
_potentially restoring previous scroll position_ makes it sound like it never was supposed to be a bulletproof solution so I guess it's okay. Maybe we could add an example to the docs/readme/changelog showing how to deal with it on the new version?

Reply:
Yeah, it was never perfect. A browser may take longer than 100ms to restore scroll position on a slow connection or long page. An addition to the docs sounds like a good idea 👍